platform/engineering/request-for-change/2019-09-30-cms-integration.md

```diff
@@ -0,0 +1,237 @@
+      S3
+  - Runs the full suite of unit, e2e, and accessibility tests
+- Content-only build
+  - Used in a content-only deploy
+  - Attempts to pull the latest CMS content
+    - Will not fall back to the cache
+  - Uses the JavaScript and CSS assets from the latest full build
+  - Runs a limited suite of tests
+- React app landing page
+  - An HTML file used as the landing page for a React application
+  - Has the normal header and footer
+  - Points to the appropriate assets for the application
+
+### Metalsmith
+Currently, `vets-website`'s build is managed by Metalsmith, a
+plugin-based static site generator. Metalsmith uses a number of custom
+plugins used to manipulate the static content in various ways. One
+such plugin fetches the content from Drupal if we pass the
+`--pull-drupal` flag to the build script (`yarn build --pull-drupal`)
+or if the local cache isn't found. Another plugin checks for broken links,
+etc.
+
+This build script is run in a number of ways:
+- The Jenkins pipeline for a full build
+- The Jenkins pipeline for a content-only build
+- Locally, with `yarn build` or `yarn watch`
+
+Separate, but related, the [preview
+server](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/platform/architecture/cms/#preview-server)
+has a modified version of the Metalsmith build (`yarn preview`). This
+is used for content editors to see how their changes show up on the
+site.
+
+Having a Metalsmith build script came from a time when `vets-website`
+had a `content/` directory from which the static pages were
+built. These days, however, all of the content comes from outside
+`vets-website`.
+
+### Content validation
+Some basic validations are performed on the content from the CMS and
+vagov-content repo.
+
+| Validation | Full build | Content-only build | Preview server |
+| ---------- | ---------- | ------------------ | -------------- |
+| Broken link checking | True; will fail the build on the `master` branch only | True; will fail the build | False |
+| Accessibility | True; will fail the build | True; will fail the build | False |
+
+The Metalsmith script performs the check for broken internal
+links. This happens before the accessibility tests in the Jenkins
+build pipelines.
+
+When broken links are discovered during a full Jenkins build, we send
+a Slack notification with a link to the build. If Jenkins was building
+the production (`master`) branch, the build will also fail, to prevent
+broken links in production.
+
+Accessibility tests run the aXe checker on all pages found in the
+sitemap. Whenever the accessibility check fails, the Jenkins build
+fails. This happens after the content build and link checker.
+
+## Design
+I propose rearchitecting the `vets-website` build to separate the
+static content building into a process distinct from the website.
+
+![separate build diagram](images/separated-content-build-proposal.png)
+
```

Comment:
One of the end states here (after validation) is "throw an error stating why the content is invalid". Do we have a mechanism to get this back to a writer that may have added a broken link? Could we possibly do this as part of a save action in Drupal?

Reply:
What we're doing now is sending a link to the broken build to Slack. The next step is sending the CSV file listing the broken links (which I already have saved in a separate file; the slack plugin just doesn't support uploading files, and I don't even know where to start investigating that one).

Ideally, this validation would be done in Drupal as part of the save action, but according to the CMS team, that's not possible. (cc @ElijahLynn @swirtSJW for confirmation.)

Reply:
Drupal can't fully handle the cases of link validation upon a save in the CMS because when creating multiple new pages, only one can be created at a time.   So it is likely that links may be added that point to a page that does not exist yet.  Requiring a link to be checked before a save could be completed would result in a very stunted workflow for the editors.

Reply:
Ah! Right. That's what it was. Thank you for that context, @swirtSJW. :raised_hands:

I don't know enough about Drupal to know if there's an alternative there, so I'll defer to @swirtSJW here and say link checking from within Drupal isn't a viable solution for now.